cvat/apps/consensus/new_intersect_merge.py

```diff
@@ -0,0 +1,1983 @@
+        img_h, img_w = self._context._item_map[dataitem_id][0].image.size
+        dist = list(segment_iou(mbbox, s, img_h=img_h, img_w=img_w) for s in cluster)
+        nearest_pos, _ = max(enumerate(dist), key=lambda e: e[1])
+        return cluster[nearest_pos]
+
+    def merge_cluster_shape(self, cluster):
+        shape = self._merge_cluster_shape_mean_box_nearest(cluster)
+        distance, _ = self._make_memoizing_distance(self.distance)
+        for ann in cluster:
+            dataset_id = self._context._item_map[self._context._ann_map[id(ann)][1]][1]
+            self._context._dataset_mean_consensus_score.setdefault(dataset_id, []).append(
+                max(0, distance(ann, shape))
+            )
+        shape_score = sum(max(0, distance(shape, s)) for s in cluster) / len(cluster)
+        return shape, shape_score
+
+    def merge_cluster(self, cluster):
+        label, label_score = self.find_cluster_label(cluster)
+
+        # when the merged annotation is rejected due to quorum constraint
+        if label is None:
+            return None
+
+        shape, shape_score = self.merge_cluster_shape(cluster)
+        shape.z_order = max(cluster, key=lambda a: a.z_order).z_order
+        shape.label = label
+        shape.attributes["score"] = label_score * shape_score if label is not None else shape_score
+
+        return shape
+
+
+@attrs
+class BboxMerger(_ShapeMerger, BboxMatcher):
+    pass
+
+
+@attrs
+class PolygonMerger(_ShapeMerger, PolygonMatcher):
+    pass
+
+
+@attrs
+class MaskMerger(_ShapeMerger, MaskMatcher):
+    pass
+
+
+@attrs
+class PointsMerger(_ShapeMerger, PointsMatcher):
+    pass
+
+
+@attrs
+class LineMerger(_ShapeMerger, LineMatcher):
+    pass
+
+
+class SkeletonMerger(_ShapeMerger, SkeletonMatcher):
+    def _merge_cluster_shape_nearest(self, cluster):
+        dist = {}
+        for idx, skeleton1 in enumerate(cluster):
+            id_skeleton1 = id(skeleton1)
+            skeleton_distance = 0
+            for skeleton2 in cluster:
+                id_skeleton2 = id(skeleton2)
+                if (id_skeleton1, id_skeleton2) in self.distances:
+                    skeleton_distance += self.distances[(id_skeleton1, id_skeleton2)]
+                elif (id_skeleton2, id_skeleton1) in self.distances:
+                    skeleton_distance += self.distances[(id_skeleton2, id_skeleton1)]
+                else:
+                    skeleton_distance += 1
+
+            dist[idx] = skeleton_distance / len(cluster)
+
+        return cluster[min(dist, key=dist.get)]
+
+    def merge_cluster_shape(self, cluster):
+        shape = self._merge_cluster_shape_nearest(cluster)
+        shape_score = sum(max(0, self.distance(shape, s)) for s in cluster) / len(cluster)
+        return shape, shape_score
+
+    # def __init__(self, categories=None):
+    #     _ShapeMerger.__init__(self)
+    #     SkeletonMatcher.__init__(self, categories)
+
+
+def match_segments(
```

Comment:
It can be imported from `quality_reports.py`

Reply:
we earlier thought not to use components from one Django app in another Django app, but rather create a file containing such common functions. So should I still use the implementation from `quality_reports.py`?

Reply:
Yes, please don't duplicate this logic.